server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -629,7 +645,9 @@ public void readFrom(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeOptionalString(type);
+        // A 7.x request allows null types but if deserialized in a 6.x node will cause nullpointer exceptions. 
+        // So we use the type accessor method here to make the type non-null (will default it to "_doc"). 
```

Comment:
Maybe add the same comment to the writeTo() above in DeleteRequest too. It helps.